staging/src/k8s.io/apimachinery/pkg/runtime/types.go

```diff
@@ -100,12 +100,14 @@ const (
 // +k8s:openapi-gen=true
 type RawExtension struct {
 	// Raw is the underlying serialization of this object.
-	//
-	// TODO: Determine how to detect ContentType and ContentEncoding of 'Raw' data.
 	Raw []byte `json:"-" protobuf:"bytes,1,opt,name=raw"`
 	// Object can hold a representation of this extension - useful for working with versioned
 	// structs.
 	Object Object `json:"-"`
+
+	// ContentType is the IANA Media Type of the bytes in Raw. The empty string is assumed to
+	// mean "application/json".
```

Comment:
When written to etcd, the only in-tree server-side uses of RawExtension certainly *hope* the content is json (though it should be validated like https://github.com/liggitt/kubernetes/commits/validate-controller-revision/)

As a general serializable struct, I don't think we can say all serializations that don't set this field mean the Raw content is JSON. RawExtension is used when serializing watch events out to responses, and contains proto content in that case, right? That means an empty ContentType proto field coming in a watch stream from an older server means the content could be proto?

Reply:
I was not really expecting to serialize this content type field, just use it for bookkeeping of how the RawExtension object was instantiated and populate it when unmarshaling json or yaml or cbor, or round-tripping to unknown, etc, so serializers would know if they needed to convert.

If unset, I think it means we don't know the content type and have to verify or heuristically detect it.

Reply:
> RawExtension is used when serializing watch events out to responses, and contains proto content in that case, right? That means an empty ContentType proto field coming in a watch stream from an older server means the content could be proto?

You're right, thanks for that example. I think, in the watch event case, the RawExtension field is guaranteed in practice to use the same encoding as the event overall. In theory, yes, the field can contain differently-encoded bytes if (and only if) the overall event encoding is Protobuf.

> If unset, I think it means we don't know the content type and have to verify or heuristically detect it.

I think I am convinced of this now. Certain RawExtension uses need to be transcodable to any supported response body encoding to avoid making those encodings second-class. That includes cases like ControllerRevision's where the RawExtension is stored opaquely. Other uses of RawExtension don't necessarily need the same guarantee, as long as the recipient knows what to do with it.